### PR TITLE
Corrected array ordering problem for the fisher module

### PR DESCRIPTION
**Pull request by AlexisMignon**
This pull request proposes the following changes
- There was a problem with the array ordering in the fisher module
- The expected input array are transposed
- The new test case directly computes the expected output and compares it to the obtained one
- Added the possibility to use input data in float64

### CODE REVIEW

File: cyvlfeat/fisher/cyfisher.pyx [hunk 1]
```diff
@@ -59,17 +83,24 @@ cpdef cy_fisher(float[:, :] X,
             n_data, n_clusters, n_dimensions, 2 * n_clusters * n_dimensions,
             square_root, normalized, fast))
 
-    cdef float[:] enc = np.zeros(2 * n_clusters * n_dimensions,
-                                 dtype=np.float32)
+    enc = np.zeros(2 * n_clusters * n_dimensions,
+                                 dtype=X.dtype)
+
+
+    data = <void*> X.data
```
Comment (patricksnape) [inline cyvlfeat/fisher/cyfisher.pyx]: I'm not keen on this syntax. Also, given [the data attribute](http://docs.scipy.org/doc/numpy-dev/release.html#assignment-of-ndarray-object-s-data-attribute) is deprecated (at least setting it), I don't think it is recommended to do this. Also, passing around `void*` is never friendly. According to the [Cython documentation](http://docs.cython.org/src/userguide/memoryviews.html#coercion-to-numpy), `np.asarray` does not copy the buffer and thus has little overhead. Also, given this [stackoverflow thread](http://stackoverflow.com/questions/18462785/what-is-the-recommended-way-of-allocating-memory-for-a-typed-memory-view) it seems like allocating memory is the least of our worries given the lack of tight loops in this code.

**Comment by patricksnape** [comment]
jenkins test this please

File: cyvlfeat/fisher/cyfisher.pyx [hunk 2]
```diff
@@ -59,24 +82,57 @@ cpdef cy_fisher(float[:, :] X,
...
+
+        encf = np.zeros(
+            2 * n_clusters * n_dimensions, dtype=X.dtype
+        )
+        meansf = MEANS
+        covarsf = COVARIANCES
+        priorsf = PRIORS
+        Xf = X
+
+        num_terms = vl_fisher_encode(&encf[0],
```
Comment (patricksnape) [inline cyvlfeat/fisher/cyfisher.pyx]: Having looked at the underyling C code I agree that this ordering is consistent given the swapping of the `n_dimensions` and `n_clusters`.

File: cyvlfeat/fisher/cyfisher.pyx [hunk 3]
```diff
@@ -20,34 +20,57 @@ from cyvlfeat._vl.fisher cimport VL_FISHER_FLAG_FAST
 
 
 @cython.boundscheck(False)
-cpdef cy_fisher(float[:, :] X,
-                float[:, :] MEANS,
-                float[:, :] COVARIANCES,
-                float[:] PRIORS,
+cpdef cy_fisher(np.ndarray X,
```
Comment (patricksnape) [inline cyvlfeat/fisher/cyfisher.pyx]: I would prefer that these were left as memory views as it allows the code to be used with systems other than Numpy. It's particularly useful if someone wants to utilise the `cpdef` and thus call this method from other Cython code without resorting to numpy.

File: cyvlfeat/fisher/tests/fisher_test.py [hunk 4]
```diff
@@ -7,48 +7,73 @@ def test_fisher_dimension():
...
-                             0.000000000, 0.000000000, 0.000000000,
-                             0.000000000, 70.519210815, 70.519210815,
-                             70.519210815, 70.519210815, 70.519210815,
-                             -0.058321182, -0.058321182, -0.058321182,
-                             -0.058321182, -0.058321182, -0.058321182,
-                             -0.058321182, -0.058321182, -0.058321182,
-                             -0.058321182, 3073.876220703, 3073.876220703,
-                             3073.876220703, 3073.876220703, 3073.876220703],
-                            dtype=np.float32)
-    assert_allclose(expected_enc, observed_enc)
```
Comment (patricksnape) [inline cyvlfeat/fisher/tests/fisher_test.py]: Would it be possible to maintain this test and just check that the input matrix transpose is still equal to this?

File: cyvlfeat/fisher/cyfisher.pyx (hunk 3 above)
Comment (AlexisMignon) [inline cyvlfeat/fisher/cyfisher.pyx]: Do you suggests that there should be two versions of the function ? cy_fisher_f and cy_fisher_d ?

**Comment by patricksnape** [comment]
I added a PR to your branch [here](https://github.com/AlexisMignon/cyvlfeat/pull/1) which shows my suggested changes to simplify the method. The complicated part was the magic `dtype_from_memoryview` method.

File: cyvlfeat/fisher/cyfisher.pyx (hunk 3 above)
Comment (patricksnape) [inline cyvlfeat/fisher/cyfisher.pyx]: I put a PR against this branch that shows my suggestion (using fused types).

Comment (patricksnape) [inline cyvlfeat/fisher/cyfisher.pyx]: ping @AlexisMignon - My [PR is here](https://github.com/alexismignon/cyvlfeat/pull/1) that gives my suggestion on how to unify this.

**Comment by patricksnape** [comment]
The merge conflict here is the `.gitignore` and it should be simple to fix.

**Comment by patricksnape** [comment]
@williamkreamer Did you ever test this - sorry - this totally slipped from my radar. If you could confirm you can still replicate your expected outputs with this that would be great. Note you may need to transpose your input matrices.

**Comment by patricksnape** [comment]
@williamkreamer Just pinging you again - if I don't hear from you soon I'm going to merge this as would like to do a release 😄

**Comment by williamkreamer** [comment]
Sorry for not responding. I haven't tested it, and don't have time to do it soon. Happy for you to merge this if you are, though

**Comment by patricksnape** [comment]
Closed by #50

**Pull request closed without merging**